### PR TITLE
query aggregated metrics with cli

### PR DESCRIPTION
**Pull request by qibobo**

### CODE REVIEW

File: src/cli/main.go [hunk 1]
```diff
@@ -65,16 +65,17 @@ OPTIONS:
...
 OPTIONS:
-	--start		Start time of metrics collected with format "yyyy-MM-ddTHH:mm:ss+/-HH:mm" or "yyyy-MM-ddTHH:mm:ssZ", default to very beginning if not specified.
-	--end		End time of the metrics collected with format "yyyy-MM-ddTHH:mm:ss+/-HH:mm" or "yyyy-MM-ddTHH:mm:ssZ", default to current time if not speficied.
-	--desc		Display in descending order, default to ascending order if not specified.
-	--output	Dump the metrics to a file in table format.
+	--app-instance-index, -i         Application instance index, default to empty so aggregated metrics will be returned.
+	--start							 Start time of metrics collected with format "yyyy-MM-ddTHH:mm:ss+/-HH:mm" or "yyyy-MM-ddTHH:mm:ssZ", default to very beginning if not specified.
+	--end							 End time of the metrics collected with format "yyyy-MM-ddTHH:mm:ss+/-HH:mm" or "yyyy-MM-ddTHH:mm:ssZ", default to current time if not speficied.
+	--desc							 Display in descending order, default to ascending order if not specified.
+	--output					     Dump the metrics to a file in table format.
```
Comment (cdlliuy) [inline src/cli/main.go]: The indents here are incorrect

**Comment by cdlliuy** [comment]
LGTM

File: src/cli/api/apihelper.go [hunk 2]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
 
 }
 
+func (helper *APIHelper) GetAggregatedMetrics(metricName string, startTime, endTime int64, desc bool, page uint64) (bool, [][]string, error) {
+
+	hasNext := false
+	if page <= 1 {
+		err := helper.CheckHealth()
+		if err != nil {
+			return hasNext, nil, err
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: return false, nil, err directly.

File: src/cli/api/apihelper.go [hunk 3]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
...
+		}
+	}
+
+	baseURL := helper.Endpoint.URL
+	queryMetricURL := strings.Replace(AggregatedMetricPath, "{appId}", helper.Client.AppId, -1)
+	queryMetricURL = strings.Replace(queryMetricURL, "{metric_type}", metricName, -1)
+	requestURL := fmt.Sprintf("%s%s", baseURL, queryMetricURL)
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: use path package to handle the path join

File: src/cli/api/apihelper.go [hunk 4]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
...
+		q.Add("order", "desc")
+	} else {
+		q.Add("order", "asc")
+	}
+	q.Add("page", strconv.FormatUint(page, 10))
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := helper.DoRequest(req)
+	if err != nil {
+		return hasNext, nil, err
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: same as above

File: src/cli/api/apihelper.go [hunk 5]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
...
+	raw, err := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != http.StatusOK {
+		var errorMsg string
+		switch resp.StatusCode {
+		case 401:
+			errorMsg = fmt.Sprintf(ui.Unauthorized, baseURL, helper.Client.CCAPIEndpoint)
+		default:
+			errorMsg, err = parseErrResponse(raw)
+			if err != nil {
+				return hasNext, nil, err
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: same as above

File: src/cli/api/apihelper.go [hunk 6]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
...
+			}
+		}
+		return hasNext, nil, errors.New(errorMsg)
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: same as above

File: src/cli/api/apihelper.go [hunk 7]
```diff
@@ -379,6 +383,77 @@ func (helper *APIHelper) GetMetrics(metricName string, startTime, endTime int64,
...
+		return hasNext, nil, err
+	}
+	defer resp.Body.Close()
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != http.StatusOK {
+		var errorMsg string
+		switch resp.StatusCode {
+		case 401:
+			errorMsg = fmt.Sprintf(ui.Unauthorized, baseURL, helper.Client.CCAPIEndpoint)
```
Comment (boyang9527) [inline src/cli/api/apihelper.go]: do we have token refresh if the access token is expired?

File: src/cli/commands/retrieve_metrics.go [hunk 8]
```diff
@@ -133,3 +148,67 @@ func RetrieveMetrics(cliConnection api.Connection, appName, metricName string, s
...
+
+	if noResult {
+		ui.SayOK()
+		ui.SayMessage(ui.AggregatedMetricsNotFound, appName)
+		ui.SayMessage(ui.UseInstanceMetricMsg)
+
+	} else {
+
+		if writer != os.Stdout {
+			ui.SayOK()
```
Comment (boyang9527) [inline src/cli/commands/retrieve_metrics.go]: why?

File: src/cli/commands/retrieve_metrics.go [hunk 9]
```diff
@@ -133,3 +148,67 @@ func RetrieveMetrics(cliConnection api.Connection, appName, metricName string, s
...
+		if err != nil {
+			return err
+		}
+
+		for _, row := range data {
+			table.Add(row)
+		}
+		if len(data) > 0 {
+			noResult = false
+			table.Print()
```
Comment (boyang9527) [inline src/cli/commands/retrieve_metrics.go]: why print table every time you got a page?

File: src/cli/api/apihelper.go (hunk 3 above)
Comment (qibobo) [reply src/cli/api/apihelper.go]: @boyang9527 
requestURL := path.Join(baseURL, queryMetricURL) will prune the first "/" of the scheme "http://", so result will be: "http:/127.0.0.1:50529/v1/apps/fakeAppId/aggregated_metric_histories/memoryused" and the request fails.

File: src/cli/api/apihelper.go (hunk 7 above)
Comment (cdlliuy) [reply src/cli/api/apihelper.go]: Token refresh happens when initializing  CF client and the application context.   Prior to the any policy/metric/history API call to autoscaler,  the cli plugin will invoke cf.GetApp() first in which the token is refreshed.

File: src/cli/commands/retrieve_metrics.go (hunk 8 above)
Comment (cdlliuy) [reply src/cli/commands/retrieve_metrics.go]: only print OK when the data is printed to a file.  Like this:
```
cf asm nodetestapp1 memoryused --start 2018-09-11T12:59:13+08:00 --end 2018-09-11T13:04:33+08:00 --output output.txt
Showing aggregated metrics for app nodetestapp1...
OK
```

File: src/cli/commands/retrieve_metrics.go (hunk 9 above)
Comment (cdlliuy) [reply src/cli/commands/retrieve_metrics.go]: after each table.Print(),   the rows are cleared  https://github.com/cloudfoundry-incubator/app-autoscaler-cli-plugin/blob/master/src/cli/ui/table.go#L52,  so table.Print() appends the pages in fact, no overlapping introduced.

File: src/cli/api/apihelper.go (hunk 3 above)
Comment (boyang9527) [reply src/cli/api/apihelper.go]: or use url package,  my concern is sometimes you passed into baseURL which ends with "/"

File: src/cli/commands/retrieve_metrics.go (hunk 9 above)
Comment (boyang9527) [reply src/cli/commands/retrieve_metrics.go]: I would prefer to add all rows to table, and then print it out

File: src/cli/commands/retrieve_metrics.go (hunk 8 above)
Comment (boyang9527) [reply src/cli/commands/retrieve_metrics.go]: this is weird ...  

1) the progressing message should be "retrieving aggregated metrics of app nodetestapp1..."
2) "OK" basically mean you successfully retrieved the metrics 
3) you probably will need to show metrics saved in file "xxx"

File: src/cli/api/apihelper.go (hunk 7 above)
Comment (boyang9527) [reply src/cli/api/apihelper.go]: got u

File: src/cli/commands/retrieve_metrics.go (hunk 9 above)
Comment (cdlliuy) [reply src/cli/commands/retrieve_metrics.go]: when we have hundred of metrics or more to return, we need to query API server several times to get all rows. In this case,  the end user may experience a long latency if we hold the print to the end.

Comment (boyang9527) [reply src/cli/commands/retrieve_metrics.go]: if you have this concern, then you probably will need to add more informational message to indicating the user you are retrieving more ....

File: src/cli/commands/retrieve_metrics.go (hunk 8 above)
Comment (cdlliuy) [reply src/cli/commands/retrieve_metrics.go]: This will be addressed in a New PR

File: src/cli/commands/retrieve_metrics.go (hunk 9 above)
Comment (cdlliuy) [reply src/cli/commands/retrieve_metrics.go]: Will reduce the metric retrieve time range in another PR.

File: src/cli/api/apihelper.go (hunk 3 above)
Comment (qibobo) [reply src/cli/api/apihelper.go]: @boyang9527 fixed in https://github.com/cloudfoundry-incubator/app-autoscaler-cli-plugin/pull/10/commits/2677630841f8ff807bb043bbf64c6a770561840d

**Comment by qibobo** [comment]
@boyang9527  updated.

**Comment by boyang9527** [comment]
LGTM
@qibobo pls resolve conflict and merge...

**Pull request merged**